### PR TITLE
Update MRI to version 2.3.6

### PR DESCRIPTION
**Pull request by aardvark179**

### CODE REVIEW

**Comment by nirvdrum** [comment]
Should this have included removing the pure Ruby JSON library and just using the one from stdlib?

**Comment by eregon** [comment]
> Should this have included removing the pure Ruby JSON library and just using the one from stdlib?

I think that's best done separately.

File: lib/mri/webrick/server.rb [hunk 1]
```diff
@@ -258,18 +258,26 @@ def run(sock)
...
-      begin
-        sock = svr.accept
-        sock.sync = true
-        Utils::set_non_blocking(sock)
-      rescue Errno::ECONNRESET, Errno::ECONNABORTED,
-             Errno::EPROTO, Errno::EINVAL
-      rescue StandardError => ex
-        msg = "#{ex.class}: #{ex.message}\n\t#{ex.backtrace[0]}"
-        @logger.error msg
+      case sock = svr.to_io.accept_nonblock(exception: false)
```
Comment (eregon) [inline lib/mri/webrick/server.rb]: Do we support `accept_nonblock`?

**Comment by chrisseaton** [comment]
>> Should this have included removing the pure Ruby JSON library and just using the one from stdlib?

> I think that's best done separately.

We'll have to create a new 2.3.6 baseline commit when we do switch to the standard JSON. But that's going to from pure Ruby to C extension isn't it? That's unlikely to be easy.

File: lib/mri/webrick/server.rb (hunk 1 above)
Comment (aardvark179) [reply lib/mri/webrick/server.rb]: We do!

Unfortunately it isn't declared as taking any arguments, so I'll need to fix that.

**Comment by chrisseaton** [comment]
Needs an entry in the change log!

File: test/truffle/ecosystem/launchers.sh [hunk 2]
```diff
@@ -20,8 +20,8 @@ cd "$(jt ruby -e 'puts Truffle::Boot.ruby_home')"
 function check_launchers() {
     if [ -n "$2" ]
     then
-        [[ "$(${1}truffleruby --version)" =~ truffleruby\ .*\ like\ ruby\ 2.3.5 ]]
-        [[ "$(${1}ruby --version)" =~ truffleruby\ .*\ like\ ruby\ 2.3.5 ]]
+        [[ "$(${1}truffleruby --version)" =~ truffleruby\ .*\ like\ ruby\ 2.3.6 ]]
+        [[ "$(${1}ruby --version)" =~ truffleruby\ .*\ like\ ruby\ 2.3.6 ]]
```
Comment (pitr-ch) [inline test/truffle/ecosystem/launchers.sh]: There is also `spec/truffle/launcher_spec.rb`

Comment (aardvark179) [reply test/truffle/ecosystem/launchers.sh]: Ah, missed that. I'd searched for `2.3.5` and `2_3_5`, but not the regexp escaped `2\.3\.5`.

Comment (aardvark179) [reply test/truffle/ecosystem/launchers.sh]: I've added notes to the docs about searching for the ruby version number, and other version numbers such as gem if they have changed.

File: doc/contributor/updating-ruby.md [hunk 3]
```diff
@@ -37,11 +37,17 @@ git commit -am 'Restore MRI modifications'
...
 * Update `doc/user/compatibility.md`
-* Update `doc/user/legal.md`
-* Search for other instances of the old version number (there are a couple in tests)
+* Update `doc/legal/legal.md`
+* Search for other instances of the old version number (there are a
+  couple in tests). The version numbers may use `.` or `_` depending
+  on context, and may be escapped as regular expressions so when
+  searching it is best to use a regexp like `a\\*[._]b\\*[._]c` to try
+  and find as many cases as possible.
+* The version numbers of cammands like `gem` may also have changed, so
```
Comment (eregon) [inline doc/contributor/updating-ruby.md]: commands*

File: lib/truffle/socket/truffle.rb [hunk 4]
```diff
@@ -95,10 +95,18 @@ def self.accept(source, new_class)
...
     end
 
-    def self.accept_nonblock(source, new_class)
+    def self.accept_nonblock(source, new_class, exception)
       source.fcntl(::Fcntl::F_SETFL, ::Fcntl::O_NONBLOCK)
 
-      accept(source, new_class)
+      begin
+        accept(source, new_class)
+      rescue Errno::EAGAIN
```
Comment (eregon) [inline lib/truffle/socket/truffle.rb]: `accept` should probably throw a IO::WaitReadable exception and not just EAGAIN, but that's out of scope.

**Pull request merged**